go-controller/pkg/ovnwebhook/podadmission.go

```diff
@@ -0,0 +1,138 @@
+package ovnwebhook
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kubevirt"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"golang.org/x/exp/maps"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	listers "k8s.io/client-go/listers/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type checkPodAnnot func(nodeLister listers.NodeLister, v annotationChange, pod *corev1.Pod, nodeName string) error
+
+// interconnectPodAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users in IC environments
+var interconnectPodAnnotations = map[string]checkPodAnnot{
+	util.OvnPodAnnotationName: func(nodeLister listers.NodeLister, v annotationChange, pod *corev1.Pod, nodeName string) error {
+		// Ignore kubevirt pods with live migration, the IP can cross node-subnet boundaries
+		if kubevirt.IsPodLiveMigratable(pod) {
+			return nil
+		}
+
+		podAnnot, err := util.UnmarshalPodAnnotation(map[string]string{util.OvnPodAnnotationName: v.value}, types.DefaultNetworkName)
+		if err != nil {
+			return err
+		}
+		node, err := nodeLister.Get(nodeName)
+		if err != nil {
+			return fmt.Errorf("could not get info on node %s from client: %w", nodeName, err)
+		}
+
+		subnets, err := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+		if err != nil {
+			return err
+		}
+		for _, ip := range podAnnot.IPs {
+			found := false
+			for _, subnet := range subnets {
+				if subnet.Contains(ip.IP) {
+					found = true
+				}
+			}
+			if !found {
+				return fmt.Errorf("%s does not belong to %s node", ip, nodeName)
+			}
+		}
+		return nil
+	},
+	util.DPUConnectionDetailsAnnot: nil,
+	util.DPUConnetionStatusAnnot:   nil,
+}
+
+type PodAdmission struct {
+	nodeLister     listers.NodeLister
+	annotations    map[string]checkPodAnnot
+	annotationKeys sets.Set[string]
+}
+
+func NewPodAdmissionWebhook(nodeLister listers.NodeLister) *PodAdmission {
+	return &PodAdmission{
+		nodeLister:     nodeLister,
+		annotations:    interconnectPodAnnotations,
+		annotationKeys: sets.New[string](maps.Keys(interconnectPodAnnotations)...),
+	}
+}
+
+func (p PodAdmission) ValidateCreate(ctx context.Context, obj runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore creation, the webhook is configured to only handle pod/status updates
```

Comment:
you shoudl handle creation as well. A user should not be able to create a pod with the annotations set. We had a bug around a user copying a current pod spec to another pod that had ovnk annotations and then create it. That should be blocked. Although I guess maybe this webhook may be restricted to just the ovnkube-node user?

Reply:
As far as I looked/tested the code, this webhook only checks the changes by ovn (i.e. 'system:ovn-node').

Reply:
I don't think we should handle pod creation in this effort. This is a separate issue and we can build on top of the work here to include that in the future. This would mean that we would need to expand the webhook to `pod` create/update events which I think is a risk I would like to initially avoid. 
Currently the only changes we verify are updates to `pod/status` and `node/status`, this means that if some user modifies annotations through `pod` resource it won't go through our webhook.

Reply:
ok lets defer for now and expand it in the future